src/webgpu/api/validation/queue/copyImageBitmapToTexture.spec.ts

```diff
@@ -65,76 +57,104 @@ interface WithMipLevel {
   mipLevel: number;
 }
 
-interface WithMipLevelSizeAndOrigin extends WithMipLevel {
-  srcImageSize: GPUExtent3DDict;
-  dstTextureSize: GPUExtent3DDict;
+interface WithSrcOrigin {
   srcOriginValue: GPUOrigin2DDict;
+}
+
+interface WithDstOriginMipLevel extends WithMipLevel {
   dstOriginValue: GPUOrigin3DDict;
 }
 
-// Helper function to generate dst origin value to take mipLevel into consideration.
+// Helper function to generate copySize for src OOB test
+function generateCopySizeForSrcOOB({ srcOriginValue }: WithSrcOrigin) {
+  const origin = {
+    x: srcOriginValue.x ? srcOriginValue.x : 0,
+    y: srcOriginValue.y ? srcOriginValue.y : 0,
+  };
+
+  // OOB origin fails even with noop copy.
+  if (origin.x > kDefaultWidth || origin.y > kDefaultHeight) {
+    return poptions('copySize', [{ width: 0, height: 0, depth: 0 }]);
+  }
+
+  const justFitCopySize = {
+    width: kDefaultWidth - origin.x,
+    height: kDefaultHeight - origin.y,
+    depth: 1,
+  };
+
+  return poptions('copySize', [
+    justFitCopySize, // correct size, maybe noop copy.
+    { width: justFitCopySize.width + 1, height: justFitCopySize.height, depth: 1 }, // OOB in width
+    { width: justFitCopySize.width, height: justFitCopySize.height + 1, depth: 1 }, // OOB in height
+  ]);
+}
+
+// Helper function to generate dst origin value based on mipLevel.
 function generateDstOriginValue({ mipLevel }: WithMipLevel) {
-  const origin = getMipMapSize(kDefaultWidth, kDefaultHeight, mipLevel);
+  const origin = computeMipMapSize(kDefaultWidth, kDefaultHeight, mipLevel);
 
   return poptions('dstOriginValue', [
     { x: 0, y: 0, z: 0 },
+    { x: origin.mipWidth - 1, y: 0, z: 0 },
+    { x: 0, y: origin.mipHeight - 1, z: 0 },
     { x: origin.mipWidth, y: 0, z: 0 },
     { x: 0, y: origin.mipHeight, z: 0 },
     { x: 0, y: 0, z: kDefaultDepth },
+    { x: origin.mipWidth + 1, y: 0, z: 0 },
+    { x: 0, y: origin.mipHeight + 1, z: 0 },
+    { x: 0, y: 0, z: kDefaultDepth + 1 },
   ]);
 }
 
-// Helper function to generate copySize to take mipLevel, origin into consideration
-function generateCopySize({
-  mipLevel,
-  srcImageSize,
-  dstTextureSize,
-  srcOriginValue,
-  dstOriginValue,
-}: WithMipLevelSizeAndOrigin) {
-  const dstMipMapSize = getMipMapSize(dstTextureSize.width, dstTextureSize.height, mipLevel);
-
-  const justFitCopySize = {
-    width:
-      dstMipMapSize.mipWidth < srcImageSize.width ? dstMipMapSize.mipWidth : srcImageSize.width,
-    height:
-      dstMipMapSize.mipHeight < srcImageSize.height ? dstMipMapSize.mipHeight : srcImageSize.height,
-    depth: kDefaultDepth,
+// Helper function to generate copySize for dst OOB test
+function generateCopySizeForDstOOB({ mipLevel, dstOriginValue }: WithDstOriginMipLevel) {
+  const dstMipMapSize = computeMipMapSize(kDefaultWidth, kDefaultHeight, mipLevel);
+  const origin = {
+    x: dstOriginValue.x ? dstOriginValue.x : 0,
```

Comment:
I'm not sure I understand why this is needed, does it happen that `dstOriginValue.x` is undefined?

Reply:
Yes. It will report the error that x can be undefined.

Reply:
@kainino0x any idea what could be done in Typescript to avoid this issue? If not LGTM

Reply:
Thanks! Works perfectly!